### PR TITLE
Add SSM Installer Policy Pack

### PR DESCRIPTION
**Pull request by SumitPopat**
image

image

image

image

image

### CODE REVIEW

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Turbot Resource ARN for the AWS Region where SSM Documents will be placed. For example: 'arn:aws::us-east-1:123456789012'
+region_resource = "<ssm_document_role_arn>"
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example:2]: Is the example correct? It doesn't seem to match up with the variable name

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Turbot Resource ARN for the AWS Region where SSM Documents will be placed. For example: 'arn:aws::us-east-1:123456789012'
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example:1]:
```suggestion
# Turbot Resource ARN for the AWS Region where SSM Documents will be placed.
# For example: 'arn:aws::us-east-1:123456789012'
```

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+## Getting Started
+
+### Requirements
+
+- [Terraform](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli)
+- Guardrails mods:
+  - [@turbot/aws-ssm](https://hub-guardrails-turbot-com-git-development-turbot.vercel.app/aws/mods/aws-ssm)
+- AWS configuration to run the stack control:
+  - S3 Bucket to store packages
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:55]:
```suggestion
  - S3 bucket to store packages
```

I don't think we should capitalize resource names, as we don't below, e.g., SSM document, A role

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+- Test edge cases
+
+## Getting Started
+
+### Requirements
+
+- [Terraform](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli)
+- Guardrails mods:
+  - [@turbot/aws-ssm](https://hub-guardrails-turbot-com-git-development-turbot.vercel.app/aws/mods/aws-ssm)
+- AWS configuration to run the stack control:
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:54]: Is it an AWS configuration or set of resources we need to run the stack control? Also, how are these resources used?

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  - S3 Bucket to store packages
+  - A role for SSM document execution with the following configured permissions:
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:56]:
```suggestion
  - An IAM role for SSM document execution with the following configured permissions:
```

We should be consistent in using `service resource` in our text

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    - SSM
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:57]: Instead we should just give a sample policy that users can copy and paste?

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+      - DescribeInstanceInformation
+      - ListCommandInvocations
+      - ListCommands
+      - SendCommand
+    - S3
+      - PutObject
+  - A role for the instance with the following configured polices:
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:64]:
```suggestion
  - An IAM role for the EC2 instance with the following configured polices:
```

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
+---
+categories: ["security"]
+primary_category: "security"
+---
+
+# Enforce Install Packages On Instances Using AWS SSM Package Installer
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:6]:
```suggestion
# Enforce Install Packages on AWS EC2 Instances Using AWS SSM Package Installer
```

I don't think `on` is capitalized in sentence case. Also, do we have any other titles where we have 2 or more services? Do we always include the cloud provider name, or just for the first service?

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+Enforcing the installation of packages on instances using the AWS SSM Package Installer is crucial for maintaining a consistent, secure, and automated software deployment process. This measure ensures that all package installations are managed and logged through AWS Systems Manager, enhancing security, compliance, and operational efficiency by providing centralized control and monitoring of software installations.
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:8]:
```suggestion
Enforcing the installation of packages on EC2 instances using the AWS SSM Package Installer is crucial for maintaining a consistent, secure, and automated software deployment process. This measure ensures that all package installations are managed and logged through AWS Systems Manager, enhancing security, compliance, and operational efficiency by providing centralized control and monitoring of software installations.
```

We should be clear on what type of instances we're talking about, since multiple services have instance types, like Lightsail.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+## Documentation
+
+- **[Review Policy settings →](https://hub-guardrails-turbot-com-git-development-turbot.vercel.app/policy-packs/enforce_install_packages_on_instances_using_ssm_package_installer/settings)**
+- Using the SSM documents
+  - There are two entry point types:
+    - Turbot-LinuxPackageInstaller-PublicNetwork
+    - Turbot-LinuxPackageInstaller-PrivateNetwork
+
+### Turbot-LinuxPackageInstaller-PublicNetwork
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:27]: Why are these under the documentation section? A follow-up question is what am I meant to do with this information?

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 8 above)
Comment (cbruno10) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md:6]: The dir name and file names should be updated too

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/variables.tf [hunk 11]
```diff
@@ -1,9 +1,9 @@
 variable "region_resource" {
-  description = "Enter the Turbot Resource ARN for the AWS Region where SSM Documents will be placed. For example: 'arn:aws::us-east-1:999999999999'"
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/variables.tf:2]: Should the var names be `region_arn` and `ssm_document_role_arn`? Do we normally give the full type of expected value in var names for policy packs?

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@SumitPopat Please see review, thanks!

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example (hunk 1 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example]: Fixed. Using `region_arn` and `ssm_document_role_arn` for variables now.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example (hunk 2 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/default.tfvars.example]: Updated.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 3 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: Updated.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 5 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: Updated

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 8 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: Updated.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 9 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: Updated.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 7 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: Updated.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 6 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: Added the same policy.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 10 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: This was added for informational purpose on what the resources would be used for once created. Felt like it was useful information and copied directly from the existing readme for this stack example.

File: policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md (hunk 8 above)
Comment (cbruno10) [reply policy_packs/aws/ssm/enforce_install_packages_on_instances_using_ssm_package_installer/README.md]: @SumitPopat Do we need `AWS` twice? I'm still not sure on that part. Do we have any other examples of multiple services in a title?

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
+---
+categories: ["security"]
+primary_category: "security"
+---
+
+# Enforce Install Packages on AWS EC2 Instances Using AWS SSM Package Installer
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:6]: Should we use `AWS` twice, or just with EC2? Using it each time seems easier to stay consistent but potentially redundant. Do we have any other examples of multiple services in a title?

Comment (cbruno10) [reply policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:6]: The dir name should be consistent too, e.g., if we do decide to keep in `AWS`, the dir name should be `on_aws_ec2_` as well (and vice versa if we only include it once).

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+      ]
+    }
+    ```
+
+  - An IAM role for the EC2 instance with the following configured permissions:
+
+    ```json
+    {
+      "Version": "2012-10-17",
+      "Statement": [
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:88]: Do we need multiple statements or could we just have 1 (since it's only 1 perm in the 2nd)?

Also, we should probably have a `Sid` for the statement.

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+- Guardrails mods:
+  - [@turbot/aws-ssm](https://hub-guardrails-turbot-com-git-development-turbot.vercel.app/aws/mods/aws-ssm)
+- AWS resources to run the stack control:
+  - S3 bucket to store packages
+  - An IAM role for SSM document execution with the following configured permissions:
+
+    ```json
+    {
+      "Version": "2012-10-17",
+      "Statement": [
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:61]: Do we need multiple statements or could we just have 1 (since it's only 1 perm in the 2nd)?

Also, we should probably have a `Sid` for the statement.

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+### Turbot-LinuxPackageInstaller-PrivateNetwork
+
+Use this entry point type when there is no direct access to the package source.
+Using this method, it is required that the URL of the package to install and a valid S3 bucket which be used to cache the package.
+
+### Turbot-LinuxS3PackageInstaller
+
+This document should never be directly run but is a helper document to install the package.
+
+### Current limitations
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:41]:
```suggestion
### Current Limitations
```

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+- Set the IAM role ARN to be used by SSM to run documents
+- Create three SSM documents in the specified region
+  - Turbot-LinuxPackageInstaller-PublicNetwork
+  - Turbot-LinuxPackageInstaller-PrivateNetwork
+  - Turbot-LinuxS3PackageInstaller
+
+## Documentation
+
+- **[Review Policy settings →](https://hub-guardrails-turbot-com-git-development-turbot.vercel.app/policy-packs/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/settings)**
+- Using the SSM documents
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:22]: I think I'm still not sure what I'd do with this info or why it's under Documentation. Documentation is normally something I'd read as reference material, but here there is none really, just our descriptions. Should we break these out into another H2 (or H3 under Getting Started somewhere) so it's more clear what these SSM docs are, where do I get them from, etc.?

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/main.tf [hunk 17]
```diff
@@ -0,0 +1,5 @@
+resource "turbot_policy_pack" "main" {
+  title       = "Enforce Install Packages on AWS EC2 Instances Using AWS SSM Package Installer"
+  description = "Ensure that all package installations are managed and logged through AWS Systems Manager, enhancing security, compliance, and operational efficiency by providing centralized control and monitoring of software installations."
+  akas        = ["aws_ssm_enforce_install_packages_on_ec2_instances_using_ssm_package_installer"]
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/main.tf:4]: Please remember to update this AKA too once we finalize the name/title in regards to cloud provider name

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@SumitPopat Please see additional review comments, thanks!

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md (hunk 12 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:6]: > Should we use AWS twice, or just with EC2? Using it each time seems easier to stay consistent but potentially redundant. Do we have any other examples of multiple services in a title?
I don't think we do. Feels redundant though. Will remove the second `AWS`.

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md (hunk 13 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:88]: Updated.

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md (hunk 14 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:61]: Updated.

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md (hunk 15 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:41]: Updated

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md (hunk 16 above)
Comment (SumitPopat) [reply policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md]: Makes sense. Add `### Using the SSM Documents` section under Getting Started.

**Review by cbruno10 — APPROVED** [review]

File: policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+            "ssm:SendCommand",
+            "s3:PutObject"
+          ],
+          "Resource": "*"
+        }
+      ]
+    }
+    ```
+
+### Using the SSM Documents
```
Comment (cbruno10) [inline policy_packs/aws/ssm/enforce_install_packages_on_ec2_instances_using_ssm_package_installer/README.md:76]: I'd suggest moving this at the bottom of the README since the documents are created once the policy pack is put into enforce mode.

**Review by cbruno10 — APPROVED** [review]
